igvm/hypervisor.py

```diff
@@ -463,21 +464,31 @@ def check_migration(self, vm, hypervisor, offline):
                 'configuration (different VLAN).'
             )
 
-    def migrate_vm(self, vm, target_hypervisor, offline, tx):
+    def migrate_vm(self, vm, target_hypervisor, offline, use_drbd, tx):
         """Migrate a VM to the given destination hypervisor"""
         self.check_migration(vm, target_hypervisor, offline)
         domain = self._get_domain(vm)
         if offline:
             target_hypervisor.create_vm_storage(vm, vm.fqdn, tx)
-            nc_listener = target_hypervisor.netcat_to_device(
-                self.vm_disk_path(vm.fqdn), tx
-            )
-            self.device_to_netcat(
-                self.vm_disk_path(domain.name()),
-                vm.dataset_obj['disk_size_gib'] * 1024**3,
-                nc_listener,
-                tx,
-            )
+            if use_drbd:
+                self.start_drbd(vm, target_hypervisor, tx=tx)
+            else:
+                if vm.is_running():
+                    vm.shutdown(tx=tx)
+                nc_listener = target_hypervisor.netcat_to_device(
+                    self.vm_disk_path(vm.fqdn), tx
+                )
+                self.device_to_netcat(
+                    self.vm_disk_path(domain.name()),
+                    vm.dataset_obj['disk_size_gib'] * 1024**3,
+                    nc_listener,
+                    tx,
+                )
+            if use_drbd:
+                self.wait_for_sync(vm, target_hypervisor, tx=tx)
+                if vm.is_running():
+                    vm.shutdown(tx=tx)
+                self.stop_drbd(vm, target_hypervisor, tx=tx)
```

Comment:
Can't you merge the two `if use_drbd` blocks?

Reply:
Maybe even better to have 2 separate functions. Please note that those things have particular order, so no, they can't be easily merged.